src/Google.Cloud.Metadata.V1/InstanceMetadata.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Collections.Generic;
+
+namespace Google.Cloud.Metadata.V1
+{
+    /// <summary>
+    /// Metadata describing the currently running Google Compute Engine VM instance.
+    /// </summary>
+    /// <seealso cref="MetadataClient.GetInstanceMetadata"/>
+    /// <seealso cref="MetadataClient.GetInstanceMetadataAsync"/>
+    public sealed class InstanceMetadata
+    {
+        // TODO: Should we change the empty string default value to be null?
+        /// <summary>
+        /// Gets the microarchitecture this machine type should use or an empty string if the machine type can be used on any
+        /// CPU platform.
+        /// </summary>
+        public string CpuPlatform { get; }
+
+        /// <summary>
+        /// Gets the custom metadata values associated with the instance.
+        /// </summary>
+        /// <seealso cref="MetadataClient.RemoveCustomInstanceMetadata"/>
+        /// <seealso cref="MetadataClient.RemoveCustomInstanceMetadataAsync"/>
+        /// <seealso cref="MetadataClient.SetCustomInstanceMetadata(string, string)"/>
+        /// <seealso cref="MetadataClient.SetCustomInstanceMetadata(IDictionary{string, string})"/>
+        /// <seealso cref="MetadataClient.SetCustomInstanceMetadataAsync(string, string, System.Threading.CancellationToken)"/>
+        /// <seealso cref="MetadataClient.SetCustomInstanceMetadataAsync(IDictionary{string, string}, System.Threading.CancellationToken)"/>
+        public IReadOnlyDictionary<string, string> CustomMetadata { get; }
+
+        /// <summary>
+        /// Gets the description of the instance which was user-supplied when created.
+        /// </summary>
+        public string Description { get; }
+
+        /// <summary>
+        /// Gets the metadata describing each disk attached to the instance.
+        /// </summary>
+        public IReadOnlyList<DiskMetadata> Disks { get; }
+
+        /// <summary>
+        /// Gets the fully qualified host name of the instance.
+        /// </summary>
+        public string HostName { get; }
+
+        /// <summary>
+        /// Gets the unique ID of the instance.
+        /// </summary>
+        public long Id { get; }
+        
+        // TODO: I'm not quite sure this comment makes sense. Look into this some more.
```

Comment:
It is the partial URL of a resource under this path: https://www.googleapis.com/compute/v1/
So for example, the MachineTypePath might be "projects/<numeric-project-id>/machineTypes/n1-standard-1". It is also used for ZonePath and NetworkInterfaceMetadata.NetworkPath. I have updated the comments on all these to be clearer. I just did some testing though and it looks like the image path is being returned as an empty string, so perhaps it is not ready to be exposed. I removed the property for now.